src/LSST/faultmgr/BasicFaultMgr.cpp

```diff
@@ -0,0 +1,157 @@
+    switch (subSystem) {
+    case SYSTEM_CONTROLLER: return "SYSTEM_CONTROLLER";
+    case FAULT_MANAGER: return "FAULT_MANAGER";
+    case POWER_SUBSYSTEM: return "POWER_SUBSYSTEM";
+    case CELL_CONTROLLER: return "CELL_CONTROLLER";
+    case TELEMETRY_LOGGER: return "TELEMETRY_LOGGER";
+    case NETWORK_INTERFACE: return "NETWORK_INTERFACE";
+    case MOTION_ENGINE: return "MOTION_ENGINE";
+    }
+    return string(" unexpected subSytem=") + to_string(subSystem);
+}
+
+BasicFaultMgr::BasicFaultMgr() {
+    // _summaryFaults, _prevFaults, and _currentFaults should all be zero already.
+    _faultEnableMask.setBitmap(FaultStatusBits::getMaskFaults()); ///< "Fault Enable Mask"
+    _defaultFaultMask.setBitmap(FaultStatusBits::getMaskFaults());; ///< "Default Fault Mask"
+    // _affectedFaultsMask, and _affectedWarnInfoMask should both be zero already.
+
+    _timeStamp = util::CLOCK::now();
+}
+
+bool BasicFaultMgr::updateFaults(BasicFaultMgr::CrioSubsystem subsystem) {
+    uint64_t diff = (_prevFaults.getBitmap() ^ _currentFaults.getBitmap()) & _faultEnableMask.getBitmap();
+    if (diff == 0) {
+        return false; // nothing needs to be done.
+    }
+    _prevFaults = _currentFaults;
+
+    // from "BasicFaultManager.lvclass:send_faults.vi"
+    // remove warning and info bits
+    uint64_t summaryAndDefault = _summaryFaults.getBitmap() & _defaultFaultMask.getBitmap();
+    // Add _currentFaults to _summaryFaults.
+    uint64_t combined = summaryAndDefault | _currentFaults.getBitmap();
+    _summaryFaults = combined;
+
+    _timeStamp = util::CLOCK::now();
+
+    return true;
+}
+
+void BasicFaultMgr::resetFaults(FaultStatusBits mask) {
+    LDEBUG("resetFaults ", mask.getAllSetBitEnums());
+    uint64_t notMask = ~(mask.getBitmap());
+    _summaryFaults.setBitmap(_summaryFaults.getBitmap() & notMask);
+    _currentFaults.setBitmap(_currentFaults.getBitmap() & notMask);
+    _prevFaults.setBitmap(_prevFaults.getBitmap() & notMask);
+
+    _timeStamp = util::CLOCK::now();
+}
+
+
+std::tuple<uint16_t, uint16_t> BasicFaultMgr::updateFaultStatus(
+        uint64_t summaryFaultStatus, uint64_t faultEnableMask,
+        uint64_t newFaultStatus, uint64_t affectedWarnInfo, uint64_t affectedFault) {
+    // affectedAll is the same as 1.MASK' from UpdateFaultStatus.vi
+    uint64_t affectedAll = affectedFault | affectedWarnInfo;
+
+    // cf is the same as CF from UpdateFaultStatus.vi
+    uint64_t cf = summaryFaultStatus;
+
+    // cfPrime is the same as 2.CF' from UpdateFaultStatus.vi
+    uint64_t cfPrime = cf & ~(affectedWarnInfo);
+
+    uint64_t newMasked = newFaultStatus & affectedAll;
+
+    // updatedSummaryFaults is the same as "Updated Summary Faults" from UpdateFaultStatus.vi
+    uint64_t updatedSummaryFaults = cf ^ (newMasked | cfPrime);
+
+    // changedBits is the same as "Changed Bits" from UpdateFaultStatus.vi
+    uint64_t changedBits = faultEnableMask & affectedAll & updatedSummaryFaults;
```

Comment:
Do you have the enough unit tests to check this part of logic is expected?

Reply:
That's a tough one. I'm basically copying the code and guessing what their intent was, which makes it really difficult to make a test that means anything. I also strongly suspect that I'll be albe to get rid of most of this as at least part of it is trying to avoid sending the same internal message over and over again, which isn't an issue as the C++ code isn't using internal message passing for faults.

Reply:
As a minimum, you could use the run the vis with different inputs to compare the results with your C++ implementation to make sure you do translate it correctly. Or you could try to write the comments of boolean calculation of this function to make sure you do understand the logic/calculation behind. I am pretty concern if you could not confidently judge the calculation details. I would suggest you create another ticket for this because the error handling is critical in the test. Thanks!